src/cmdlib.sh

```diff
@@ -127,10 +144,105 @@ EOF
     fi
 
     rm -f ${changed_stamp}
-    set -x
-    sudo rpm-ostree compose tree --repo=${workdir}/repo-build --cachedir=${workdir}/cache \
-         --touch-if-changed "${changed_stamp}" \
-         ${treecompose_args} \
-         ${TREECOMPOSE_FLAGS:-} ${manifest} "$@"
-    set +x
+
+    # really, we don't need to do this whole jazz for `fetch`, but meh, it's
+    # just some minimal I/O in that case
+    cat > ${TMPDIR}/compose.sh <<EOF
+SUDO=
+if [ \$(id -u) != 0 ]; then
+    SUDO=sudo
+fi
+
+# build in a temp bare-user repo in cache/ so we can hardlink from pkgcache
+\$SUDO ostree --repo=${workdir}/cache/repo-build init --mode=bare-user
+
+# seed with commit metadata of last build
+# TODO: don't rely on ${ref}
+if ostree --repo=${workdir}/repo rev-parse ${ref} 2>/dev/null; then
+    \$SUDO ostree --repo=${workdir}/cache/repo-build pull-local \
+        ${workdir}/repo ${ref} --commit-metadata-only
+fi
+
+\$SUDO rpm-ostree compose tree --repo=${workdir}/cache/repo-build \
+    --cachedir=${workdir}/cache --touch-if-changed "${changed_stamp}" \
+    ${treecompose_args} ${TREECOMPOSE_FLAGS:-} \
+    ${manifest} $@
+
+\$SUDO chmod -R a+rX ${workdir}/cache/repo-build/objects
+# can't use ${changed_stamp}; it has different semantics with --download-only
+if ostree --repo=${workdir}/cache/repo-build rev-parse ${ref} 2>/dev/null; then
+    ostree --repo=${workdir}/repo pull-local ${workdir}/cache/repo-build ${ref}
+fi
+
+# just nuke it instead of pruning, it's fast to rebuild from pkgcache hardlinks
+\$SUDO rm -rf ${workdir}/repo-build
+EOF
+
+    # this is the heart of the privs vs no privs dual path
+    if has_privileges; then
+        . ${TMPDIR}/compose.sh
+    else
+        runvm bash ${TMPDIR}/compose.sh
+    fi
+}
+
+runvm() {
+    local vmpreparedir=${workdir}/tmp/supermin.prepare
+    local vmbuilddir=${workdir}/tmp/supermin.build
+
+    # use REBUILDVM=1 if e.g. hacking on rpm-ostree/ostree and wanting to get
```

Comment:
the problem with not building the vm every time is that it can get out of date (i.e. new deps needed, etc). This is probably fine for your local dev case, but not ideal for having to go do something in your CI. I converted my master branch to just rebuild it every time. I found it didn't really add that much extra time to the compose.

Reply:
then again, if we are doing it for fetch and build then it starts to get a bit more heavyweight

Reply:
I think we could make fetch unprivileged without too much effort BTW.

Reply:
> I think we could make fetch unprivileged without too much effort BTW.

right, but if we are in "unpriv mode" we still need to run it in the VM so it can access the cache I think.